File: mlir/circt/frontends/PyCDE/src/devicedb.py
#  Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
#  See https://llvm.org/LICENSE.txt for license information.
#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

from __future__ import annotations
from typing import Any, List, Optional, Tuple, Union

from .support import get_user_loc

from .circt.dialects import msft
from .circt.support import attribute_to_var
from .circt.ir import Attribute, StringAttr, ArrayAttr, FlatSymbolRefAttr

from functools import singledispatchmethod

PrimitiveType = msft.PrimitiveType


class PhysLocation:
  __slots__ = ["_loc"]

  @singledispatchmethod
  def __init__(self,
               prim_type: Union[str, PrimitiveType],
               x: int,
               y: int,
               num: Optional[int] = 0):

    if isinstance(prim_type, str):
      prim_type = getattr(PrimitiveType, prim_type)
    # TODO: Once we get into non-zero num primitives, this needs to be updated.
    if num is None:
      num = 0

    assert isinstance(prim_type, PrimitiveType)
    assert isinstance(x, int)
    assert isinstance(y, int)
    assert isinstance(num, int)
    self._loc = msft.PhysLocationAttr.get(prim_type, x, y, num)

  @__init__.register(msft.PhysLocationAttr)
  def __from_loc(self, loc):
    self._loc = loc

  @__init__.register(Attribute)
  def __from_attr(self, loc):
    self._loc = msft.PhysLocationAttr(loc)

  def __str__(self) -> str:
    loc = self._loc
    return f"PhysLocation<{loc.devtype}, x:{loc.x}, y:{loc.y}, num:{loc.num}>"

  def __repr__(self) -> str:
    return self.__str__()


class LocationVector:
  from .types import Type

  __slots__ = ["type", "_loc"]

  @singledispatchmethod
  def __init__(self, type: Type, locs: List[Optional[Tuple[int, int, int]]]):
    assert len(locs) == type.bitwidth, \
      "List length must match reg bitwidth"
    from .circt.dialects import msft as circt_msft
    self.type = type
    phys_locs: List[circt_msft.PhysLocationAttr] = list()
    for loc in locs:
      if loc is None:
        phys_locs.append(None)
      else:
        phys_locs.append(
            circt_msft.PhysLocationAttr.get(PrimitiveType.FF, loc[0], loc[1],
                                            loc[2]))
    self._loc = circt_msft.LocationVectorAttr.get(type._type, phys_locs)

  @__init__.register(msft.LocationVectorAttr)
  def __from_loc(self, loc):
    self._loc = loc

  @property
  def locs(self) -> List[PhysLocation]:
    return [PhysLocation(loc) if loc is not None else None for loc in self._loc]

  def __str__(self) -> str:
    locs = [f"{loc}" for loc in self.locs]
    return f"LocationVector<{self.type}, [" + ", ".join(locs) + "]>"

  def __repr__(self) -> str:
    return self.__str__()


class PhysicalRegion:
  _counter = 0
  _used_names = set([])

  __slots__ = ["_physical_region"]

  def __init__(self, name: str = None, bounds: list = None):
    if name is None or name in PhysicalRegion._used_names:
      prefix = name if name is not None else "region"
      name = f"{prefix}_{PhysicalRegion._counter}"
      while name in PhysicalRegion._used_names:
        PhysicalRegion._counter += 1
        name = f"{prefix}_{PhysicalRegion._counter}"
    PhysicalRegion._used_names.add(name)

    if bounds is None:
      bounds = []

    name_attr = StringAttr.get(name)
    bounds_attr = ArrayAttr.get(bounds)
    self._physical_region = msft.PhysicalRegionOp(name_attr, bounds_attr)

  def add_bounds(self, x_bounds: tuple, y_bounds: tuple):
    """Add a new bounding box to the region."""
    if (len(x_bounds) != 2):
      raise ValueError(f"expected lower and upper x bounds, got: {x_bounds}")
    if (len(y_bounds) != 2):
      raise ValueError(f"expected lower and upper y bounds, got: {y_bounds}")

    x_min, x_max = x_bounds
    y_min, y_max = y_bounds
    bounds = msft.PhysicalBoundsAttr.get(x_min, x_max, y_min, y_max)

    self._physical_region.add_bounds(bounds)

    return self

  def get_ref(self):
    """Get a pair suitable for add_attribute to attach to an operation."""
    name = self._physical_region.sym_name.value
    return ("loc", msft.PhysicalRegionRefAttr.get(name))


class PrimitiveDB:
  __slots__ = ["_db"]

  def __init__(self):
    self._db = msft.PrimitiveDB()

  def add_coords(self,
                 prim_type: Union[str, PrimitiveType],
                 x: int,
                 y: int,
                 num: Optional[int] = None):
    self.add(PhysLocation(prim_type, x, y, num))

  def add(self, physloc: PhysLocation):
    self._db.add_primitive(physloc._loc)


class PlacementDB:
  from .instance import Instance

  __slots__ = ["_db", "_sys"]

  def __init__(self, sys, _circt_mod, seed: Optional[PrimitiveDB]):
    self._db = msft.PlacementDB(_circt_mod, seed._db if seed else None)
    self._sys = sys

  def get_instance_at(self, loc: PhysLocation):
    """Get the instance placed at `loc`. Returns (Instance, subpath)."""
    loc_op = self._db.get_instance_at(loc._loc)
    if loc_op is None:
      return None
    inst = self._sys._op_cache.get_or_create_inst_from_op(loc_op.parent.opview)
    subpath = attribute_to_var(loc_op.opview.subPath)
    return (inst, subpath)

  def reserve_location(self, loc: PhysLocation, entity: EntityExtern):
    sym_name = entity._entity_extern.sym_name.value
    ref = FlatSymbolRefAttr.get(sym_name)
    path = ArrayAttr.get([ref])
    subpath = ""
    self._db.add_placement(loc._loc, path, subpath, entity._entity_extern)

  def place(self, inst: Instance, loc: PhysLocation, subpath: str = ""):
    self._db.place(inst._dyn_inst.operation, loc._loc, subpath, get_user_loc())


class EntityExtern:
  __slots__ = ["_entity_extern"]

  def __init__(self, tag: str, metadata: Any = ""):
    self._entity_extern = msft.EntityExternOp.create(tag, metadata)

File: mlir/circt/lib/Dialect/MSFT/DeviceDB.cpp
//===- DeviceDB.cpp - Implement a device database -------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//

#include "circt/Dialect/MSFT/DeviceDB.h"
#include "circt/Dialect/HW/HWAttributes.h"
#include "circt/Dialect/MSFT/MSFTOps.h"

#include "mlir/IR/Builders.h"
#include "llvm/ADT/TypeSwitch.h"

using namespace circt;
using namespace msft;

//===----------------------------------------------------------------------===//
// PrimitiveDB.
//===----------------------------------------------------------------------===//
// NOTE: Nothing in this implementation is in any way the most optimal
// implementation. We put off deciding what the correct data structure is until
// we have a better handle of the operations it must accelerate. Performance is
// not an immediate goal.
//===----------------------------------------------------------------------===//

PrimitiveDB::PrimitiveDB(MLIRContext *ctxt) : ctxt(ctxt) {}

/// Assign an instance to a primitive. Return false if another instance is
/// already placed at that location.
LogicalResult PrimitiveDB::addPrimitive(PhysLocationAttr loc) {
  DenseSet<PrimitiveType> &primsAtLoc = getLeaf(loc);
  PrimitiveType prim = loc.getPrimitiveType().getValue();
  if (primsAtLoc.contains(prim))
    return failure();
  primsAtLoc.insert(prim);
  return success();
}

/// Assign an instance to a primitive. Return false if another instance is
/// already placed at that location.
/// Check to see if a primitive exists.
bool PrimitiveDB::isValidLocation(PhysLocationAttr loc) {
  DenseSet<PrimitiveType> primsAtLoc = getLeaf(loc);
  return primsAtLoc.contains(loc.getPrimitiveType().getValue());
}

PrimitiveDB::DimPrimitiveType &PrimitiveDB::getLeaf(PhysLocationAttr loc) {
  return placements[loc.getX()][loc.getY()][loc.getNum()];
}

void PrimitiveDB::foreach (
    function_ref<void(PhysLocationAttr)> callback) const {
  for (const auto &x : placements)
    for (const auto &y : x.second)
      for (const auto &n : y.second)
        for (auto p : n.second)
          callback(PhysLocationAttr::get(ctxt, PrimitiveTypeAttr::get(ctxt, p),
                                         x.first, y.first, n.first));
}

//===----------------------------------------------------------------------===//
// PlacementDB.
//===----------------------------------------------------------------------===//
// NOTE: Nothing in this implementation is in any way the most optimal
// implementation. We put off deciding what the correct data structure is until
// we have a better handle of the operations it must accelerate. Performance is
// not an immediate goal.
//===----------------------------------------------------------------------===//

PlacementDB::PlacementDB(mlir::ModuleOp topMod)
    : ctxt(topMod->getContext()), topMod(topMod), seeded(false) {
  addDesignPlacements();
}
PlacementDB::PlacementDB(mlir::ModuleOp topMod, const PrimitiveDB &seed)
    : ctxt(topMod->getContext()), topMod(topMod), seeded(false) {

  seed.foreach ([this](PhysLocationAttr loc) { (void)getLeaf(loc); });
  seeded = true;
  addDesignPlacements();
}

/// Assign an instance to a primitive. Return null if another instance is
/// already placed at that location
PDPhysLocationOp PlacementDB::place(DynamicInstanceOp inst,
                                    PhysLocationAttr loc, StringRef subPath,
                                    Location srcLoc) {
  StringAttr subPathAttr;
  if (!subPath.empty())
    subPathAttr = StringAttr::get(inst->getContext(), subPath);
  PDPhysLocationOp locOp =
      OpBuilder(inst.getBody())
          .create<PDPhysLocationOp>(srcLoc, loc, subPathAttr,
                                    FlatSymbolRefAttr());
  if (succeeded(insertPlacement(locOp, locOp.getLoc())))
    return locOp;
  locOp->erase();
  return {};
}
PDRegPhysLocationOp PlacementDB::place(DynamicInstanceOp inst,
                                       LocationVectorAttr locs,
                                       Location srcLoc) {
  PDRegPhysLocationOp locOp =
      OpBuilder(inst.getBody())
          .create<PDRegPhysLocationOp>(srcLoc, locs, FlatSymbolRefAttr());
  for (PhysLocationAttr loc : locs.getLocs())
    if (failed(insertPlacement(locOp, loc))) {
      locOp->erase();
      return {};
    }
  return locOp;
}

LogicalResult PlacementDB::insertPlacement(DynInstDataOpInterface op,
                                           PhysLocationAttr loc) {
  if (!loc)
    return success();
  PlacementCell *leaf = getLeaf(loc);
  if (!leaf)
    return op->emitOpError("Could not apply placement. Invalid location: ")
           << loc;
  if (leaf->locOp != nullptr)
    return op->emitOpError("Could not apply placement ")
           << loc << ". Position already occupied by "
           << cast<DynamicInstanceOp>(leaf->locOp->getParentOp())
                  .globalRefPath();

  leaf->locOp = op;
  return success();
}

/// Assign an operation to a physical region. Return false on failure.
PDPhysRegionOp PlacementDB::placeIn(DynamicInstanceOp inst,
                                    DeclPhysicalRegionOp physregion,
                                    StringRef subPath, Location srcLoc) {
  StringAttr subPathAttr;
  if (!subPath.empty())
    subPathAttr = StringAttr::get(inst->getContext(), subPath);
  PDPhysRegionOp regOp =
      OpBuilder::atBlockEnd(&inst.getBody().front())
          .create<PDPhysRegionOp>(srcLoc, FlatSymbolRefAttr::get(physregion),
                                  subPathAttr, FlatSymbolRefAttr());
  regionPlacements.push_back(regOp);
  return regOp;
}

/// Using the operation attributes, add the proper placements to the database.
/// Return the number of placements which weren't added due to conflicts.
size_t PlacementDB::addPlacements(DynamicInstanceOp inst) {
  size_t numFailed = 0;
  inst->walk([&](Operation *op) {
    LogicalResult added = TypeSwitch<Operation *, LogicalResult>(op)
                              .Case([&](PDPhysLocationOp op) {
                                return insertPlacement(op, op.getLoc());
                              })
                              .Case([&](PDRegPhysLocationOp op) {
                                ArrayRef<PhysLocationAttr> locs =
                                    op.getLocs().getLocs();
                                for (auto loc : locs)
                                  if (failed(insertPlacement(op, loc)))
                                    return failure();
                                return success();
                              })
                              .Case([&](PDPhysRegionOp op) {
                                regionPlacements.push_back(op);
                                return success();
                              })
                              .Default([](Operation *op) { return failure(); });
    if (failed(added))
      ++numFailed;
  });

  return numFailed;
}

/// Walk the entire design adding placements.
size_t PlacementDB::addDesignPlacements() {
  size_t failed = 0;
  for (auto inst : topMod.getOps<DynamicInstanceOp>())
    failed += addPlacements(inst);
  return failed;
}

/// Remove the placement at a given location. Returns failure if nothing was
/// placed there.
void PlacementDB::removePlacement(PDPhysLocationOp locOp) {
  removePlacement(locOp, locOp.getLoc());
  locOp.erase();
}

/// Move the placement at a given location to a new location. Returns failure
/// if nothing was placed at the previous location or something is already
/// placed at the new location.
LogicalResult PlacementDB::movePlacement(PDPhysLocationOp locOp,
                                         PhysLocationAttr newLoc) {
  PhysLocationAttr from = locOp.getLoc();
  if (failed(movePlacementCheck(locOp, from, newLoc)))
    return failure();
  locOp.setLocAttr(newLoc);
  movePlacement(locOp, from, newLoc);
  return success();
}

/// Remove the placement at a given location. Returns failure if nothing was
/// placed there.
void PlacementDB::removePlacement(PDRegPhysLocationOp locOp) {
  for (PhysLocationAttr loc : locOp.getLocs().getLocs())
    if (loc)
      removePlacement(locOp, loc);
  locOp.erase();
}

/// Move the placement at a given location to a new location. Returns failure
/// if nothing was placed at the previous location or something is already
/// placed at the new location.
LogicalResult PlacementDB::movePlacement(PDRegPhysLocationOp locOp,
                                         LocationVectorAttr newLocs) {
  ArrayRef<PhysLocationAttr> fromLocs = locOp.getLocs().getLocs();

  // Check that each move/insert/delete will succeed before doing any of the
  // mutations.
  for (auto [from, to] : llvm::zip(fromLocs, newLocs.getLocs())) {
    // If 'from' and 'to' are both non-null, this location is being moved.
    if (from && to && failed(movePlacementCheck(locOp, from, to)))
      return failure();
    // If only 'to' is valid, this location is the initial placement.
    if (to && getInstanceAt(to))
      return failure();
    // If 'to' isn't valid, it's a placement removal which will always succeed.
  }

  // Mutate.
  for (auto [from, to] : llvm::zip(fromLocs, newLocs.getLocs())) {
    // If 'from' and 'to' are both non-null, this location is being moved.
    if (from && to)
      movePlacement(locOp, from, to);
    // If 'to' isn't valid, it's a placement removal.
    else if (from)
      removePlacement(locOp, from);
    // If only 'to' is valid, this location is the initial placement. Since we
    // checked that there isn't anything currently located at 'to' this call
    // will never fail.
    else if (to)
      (void)insertPlacement(locOp, to);
  }

  locOp.setLocsAttr(newLocs);
  return success();
}

void PlacementDB::removePlacement(DynInstDataOpInterface op,
                                  PhysLocationAttr loc) {
  PlacementCell *leaf = getLeaf(loc);
  assert(leaf && "Could not find op at location specified by op");
  assert(leaf->locOp == op);
  leaf->locOp = {};
}

LogicalResult PlacementDB::movePlacementCheck(DynInstDataOpInterface op,
                                              PhysLocationAttr from,
                                              PhysLocationAttr to) {
  if (from == to)
    return success();

  PlacementCell *oldLeaf = getLeaf(from);
  PlacementCell *newLeaf = getLeaf(to);

  if (!oldLeaf || !newLeaf)
    return failure();

  if (oldLeaf->locOp == nullptr)
    return op.emitError("cannot move from a location not occupied by "
                        "specified op. Currently unoccupied");
  if (oldLeaf->locOp != op)
    return op.emitError("cannot move from a location not occupied by "
                        "specified op. Currently occupied by ")
           << oldLeaf->locOp;
  if (newLeaf->locOp)
    return op.emitError(
               "cannot move to new location since location is occupied by ")
           << cast<DynamicInstanceOp>(newLeaf->locOp->getParentOp())
                  .globalRefPath();
  return success();
}

void PlacementDB::movePlacement(DynInstDataOpInterface op,
                                PhysLocationAttr from, PhysLocationAttr to) {
  assert(succeeded(movePlacementCheck(op, from, to)) &&
         "Call `movePlacementCheck` first to ensure that move is legal.");
  if (from == to)
    return;
  PlacementCell *oldLeaf = getLeaf(from);
  PlacementCell *newLeaf = getLeaf(to);
  newLeaf->locOp = op;
  oldLeaf->locOp = {};
}

/// Lookup the instance at a particular location.
DynInstDataOpInterface PlacementDB::getInstanceAt(PhysLocationAttr loc) {
  auto innerMap = placements[loc.getX()][loc.getY()][loc.getNum()];
  auto instF = innerMap.find(loc.getPrimitiveType().getValue());
  if (instF == innerMap.end())
    return {};
  if (!instF->getSecond().locOp)
    return {};
  return instF->getSecond().locOp;
}

PhysLocationAttr PlacementDB::getNearestFreeInColumn(PrimitiveType prim,
                                                     uint64_t columnNum,
                                                     uint64_t nearestToY) {
  // Simplest possible algorithm.
  PhysLocationAttr nearest = {};
  walkPlacements(
      [&nearest, nearestToY](PhysLocationAttr loc, Operation *locOp) {
        if (locOp)
          return;
        if (!nearest) {
          nearest = loc;
          return;
        }
        int64_t curDist =
            std::abs((int64_t)nearestToY - (int64_t)nearest.getY());
        int64_t replDist = std::abs((int64_t)nearestToY - (int64_t)loc.getY());
        if (replDist < curDist)
          nearest = loc;
      },
      std::make_tuple(columnNum, columnNum, -1, -1), prim);
  return nearest;
}

PlacementDB::PlacementCell *PlacementDB::getLeaf(PhysLocationAttr loc) {
  PrimitiveType primType = loc.getPrimitiveType().getValue();

  DimNumMap &nums = placements[loc.getX()][loc.getY()];
  if (!seeded)
    return &nums[loc.getNum()][primType];
  if (!nums.count(loc.getNum()))
    return {};

  DimDevType &primitives = nums[loc.getNum()];
  if (primitives.count(primType) == 0)
    return {};
  return &primitives[primType];
}

/// Walker for placements.
void PlacementDB::walkPlacements(
    function_ref<void(PhysLocationAttr, DynInstDataOpInterface)> callback,
    std::tuple<int64_t, int64_t, int64_t, int64_t> bounds,
    std::optional<PrimitiveType> primType, std::optional<WalkOrder> walkOrder) {
  uint64_t xmin = std::get<0>(bounds) < 0 ? 0 : std::get<0>(bounds);
  uint64_t xmax = std::get<1>(bounds) < 0 ? std::numeric_limits<uint64_t>::max()
                                          : (uint64_t)std::get<1>(bounds);
  uint64_t ymin = std::get<2>(bounds) < 0 ? 0 : std::get<2>(bounds);
  uint64_t ymax = std::get<3>(bounds) < 0 ? std::numeric_limits<uint64_t>::max()
                                          : (uint64_t)std::get<3>(bounds);

  // TODO: Since the data structures we're using aren't sorted, the best we can
  // do is iterate and filter. If a specific order is requested, we sort the
  // keys by that as we go. Once we get to performance, we'll figure out the
  // right data structure.

  auto maybeSort = [](auto &container, auto direction) {
    if (!direction.has_value())
      return;
    if (*direction == Direction::NONE)
      return;

    llvm::sort(container, [direction](auto colA, auto colB) {
      if (*direction == Direction::ASC)
        return colA.first < colB.first;

      return colA.first > colB.first;
    });
  };

  // X loop.
  SmallVector<std::pair<size_t, DimYMap>> cols(placements.begin(),
                                               placements.end());
  maybeSort(cols, llvm::transformOptional(walkOrder,
                                          [](auto wo) { return wo.columns; }));
  for (const auto &colF : cols) {
    size_t x = colF.first;
    if (x < xmin || x > xmax)
      continue;
    DimYMap yMap = colF.second;

    // Y loop.
    SmallVector<std::pair<size_t, DimNumMap>> rows(yMap.begin(), yMap.end());
    maybeSort(rows, llvm::transformOptional(walkOrder,
                                            [](auto wo) { return wo.rows; }));
    for (const auto &rowF : rows) {
      size_t y = rowF.first;
      if (y < ymin || y > ymax)
        continue;
      DimNumMap numMap = rowF.second;

      // Num loop.
      for (auto &numF : numMap) {
        size_t num = numF.getFirst();
        DimDevType devMap = numF.getSecond();

        // DevType loop.
        for (auto &devF : devMap) {
          PrimitiveType devtype = devF.getFirst();
          if (primType && devtype != *primType)
            continue;
          PlacementCell &inst = devF.getSecond();

          // Marshall and run the callback.
          PhysLocationAttr loc = PhysLocationAttr::get(
              ctxt, PrimitiveTypeAttr::get(ctxt, devtype), x, y, num);
          callback(loc, inst.locOp);
        }
      }
    }
  }
}

/// Walk the region placement information.
void PlacementDB::walkRegionPlacements(
    function_ref<void(PDPhysRegionOp)> callback) {
  for (auto regOp : regionPlacements)
    callback(regOp);
}
